File: disha/core/mesh.py
"""Disha offline mesh: signed envelopes + multi-hop relay (inspired by MeshGemma).

Adds the two things our mesh design was missing versus MeshGemma:
  * Cryptographically SIGNED incident envelopes (Ed25519), so peers verify a report
    before trusting it — critical against misinformation in a disaster.
  * Lamport-clock ordering + per-node dedup for robust multi-hop flooding.

Signing is pluggable:
  * Ed25519Signer - real public-key signatures via `cryptography` (lazy import); this is
    what runs on-device (mirrors MeshGemma's tweetnacl/Ed25519).
  * DevSigner     - stdlib-only, deterministic, NON-cryptographic stand-in so the envelope /
                    relay / dedup logic is testable offline without any crypto dependency.

Transport (Nearby Connections on Android) is intentionally NOT here — this module is the
signed-envelope + relay logic that sits on top of whatever transport carries the bytes.
"""
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from typing import Optional


# --------------------------------------------------------------------------- #
# Canonical serialization (both sides must hash identical bytes)
# --------------------------------------------------------------------------- #
def canonical_bytes(obj: dict) -> bytes:
    """Deterministic JSON: sorted keys, no whitespace."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=False).encode("utf-8")


# --------------------------------------------------------------------------- #
# Signers
# --------------------------------------------------------------------------- #
class Signer:
    """node_id (public identity), sign(bytes)->hex, verify(node_id, bytes, hex)->bool."""
    node_id: str

    def sign(self, data: bytes) -> str:
        raise NotImplementedError

    @staticmethod
    def verify(node_id: str, data: bytes, sig_hex: str) -> bool:
        raise NotImplementedError


class DevSigner(Signer):
    """Deterministic, NON-cryptographic stand-in for offline tests/CI.

    sig = sha256(node_id || data). Verifiable by anyone (no secret) — good enough to exercise
    envelope/relay/dedup + tamper detection, NOT for production. Use Ed25519Signer on device.
    """

    def __init__(self, node_id: str):
        self.node_id = node_id

    @staticmethod
    def _digest(node_id: str, data: bytes) -> str:
        return hashlib.sha256(node_id.encode() + b"|" + data).hexdigest()

    def sign(self, data: bytes) -> str:
        return self._digest(self.node_id, data)

    @staticmethod
    def verify(node_id: str, data: bytes, sig_hex: str) -> bool:
        return DevSigner._digest(node_id, data) == sig_hex


class Ed25519Signer(Signer):
    """Real Ed25519 signatures (on-device). Lazy import so the module loads without crypto."""

    def __init__(self, node_id: Optional[str] = None, private_bytes: Optional[bytes] = None):
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
        self._sk = (Ed25519PrivateKey.from_private_bytes(private_bytes)
                    if private_bytes else Ed25519PrivateKey.generate())
        from cryptography.hazmat.primitives import serialization
        pub = self._sk.public_key().public_bytes(
            serialization.Encoding.Raw, serialization.PublicFormat.Raw)
        self.node_id = node_id or pub.hex()   # public key IS the identity

    def sign(self, data: bytes) -> str:
        return self._sk.sign(data).hex()

    @staticmethod
    def verify(node_id: str, data: bytes, sig_hex: str) -> bool:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature
        try:
            Ed25519PublicKey.from_public_bytes(bytes.fromhex(node_id)).verify(
                bytes.fromhex(sig_hex), data)
            return True
        except (InvalidSignature, ValueError):
            return False


# --------------------------------------------------------------------------- #
# Signed envelope
# --------------------------------------------------------------------------- #
@dataclass
class SignedEnvelope:
    sender: str            # node_id / public key
    msg_id: str
    lamport: int
    payload: dict          # e.g. an SOSReport dict
    sig: str
    type: str = "sos"
    version: int = 1
    ttl: int = 4           # mutable relay metadata — NOT part of the signed content

    def signed_content(self) -> dict:
        return {"version": self.version, "sender": self.sender, "msg_id": self.msg_id,
                "lamport": self.lamport, "type": self.type, "payload": self.payload}

    @classmethod
    def create(cls, signer: Signer, payload: dict, msg_id: str, lamport: int,
               type: str = "sos", ttl: int = 4) -> "SignedEnvelope":
        env = cls(sender=signer.node_id, msg_id=msg_id, lamport=lamport, payload=payload,
                  sig="", type=type, ttl=ttl)
        env.sig = signer.sign(canonical_bytes(env.signed_content()))
        return env

    def verify(self, signer_cls=DevSigner) -> bool:
        return signer_cls.verify(self.sender, canonical_bytes(self.signed_content()), self.sig)

    def to_dict(self) -> dict:
        d = self.signed_content()
        d.update({"sig": self.sig, "ttl": self.ttl})
        return d


# --------------------------------------------------------------------------- #
# Node: Lamport clock + dedup + multi-hop flooding
# --------------------------------------------------------------------------- #
@dataclass
class MeshNode:
    node_id: str
    signer_cls: type = DevSigner
    clock: int = 0
    seen: set = field(default_factory=set)
    inbox: list = field(default_factory=list)     # accepted payloads

    def tick(self) -> int:
        self.clock += 1
        return self.clock

    def receive(self, env: SignedEnvelope):
        """Return an envelope to re-broadcast (ttl-1), or None (dropped/duplicate/invalid)."""
        if not env.verify(self.signer_cls):
            return None                            # tamper/forgery -> drop
        if env.msg_id in self.seen:
            return None                            # duplicate -> drop
        self.seen.add(env.msg_id)
        self.clock = max(self.clock, env.lamport) + 1   # Lamport update
        self.inbox.append(env.payload)
        if env.ttl > 0:
            fwd = SignedEnvelope(**{**env.__dict__})
            fwd.ttl = env.ttl - 1                  # signature still valid (ttl not signed)
            return fwd
        return None

File: disha-android/app/src/main/java/com/example/gemmachat/core/Mesh.kt
package com.example.gemmachat.core

import java.security.MessageDigest

/**
 * Offline mesh: signed envelopes + multi-hop relay — Kotlin port of disha/core/mesh.py
 * (inspired by MeshGemma). Signing is pluggable; DevSigner (SHA-256) is a testable stand-in,
 * production uses Ed25519 (Android java.security, API 33+) — swap [Signer] implementation.
 */

/** Deterministic canonical JSON (sorted keys, compact) so both sides hash identical bytes. */
object Canonical {
    fun of(value: Any?): String = when (value) {
        null -> "null"
        is String -> "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\""
        is Boolean -> value.toString()
        is Int -> value.toString()
        is Long -> value.toString()
        is Double -> if (value % 1.0 == 0.0) value.toLong().toString() else value.toString()
        is Map<*, *> -> value.entries.sortedBy { it.key.toString() }
            .joinToString(",", "{", "}") { "\"${it.key}\":${of(it.value)}" }
        is List<*> -> value.joinToString(",", "[", "]") { of(it) }
        else -> "\"$value\""
    }
    fun bytes(value: Any?): ByteArray = of(value).toByteArray(Charsets.UTF_8)
}

interface Signer {
    val nodeId: String
    fun sign(data: ByteArray): String
}

/** Deterministic NON-cryptographic stand-in (SHA-256) for tests/CI. Not for production. */
class DevSigner(override val nodeId: String) : Signer {
    override fun sign(data: ByteArray): String = digest(nodeId, data)
    companion object {
        fun digest(nodeId: String, data: ByteArray): String {
            val md = MessageDigest.getInstance("SHA-256")
            md.update(nodeId.toByteArray(Charsets.UTF_8)); md.update('|'.code.toByte()); md.update(data)
            return md.digest().joinToString("") { "%02x".format(it) }
        }
        fun verify(nodeId: String, data: ByteArray, sigHex: String): Boolean =
            digest(nodeId, data) == sigHex
    }
}

data class SignedEnvelope(
    val sender: String,
    val msgId: String,
    val lamport: Int,
    val payload: Map<String, Any?>,
    var sig: String = "",
    val type: String = "sos",
    val version: Int = 1,
    var ttl: Int = 4,        // relay metadata — NOT part of the signed content
) {
    fun signedContent(): Map<String, Any?> = mapOf(
        "version" to version, "sender" to sender, "msg_id" to msgId,
        "lamport" to lamport, "type" to type, "payload" to payload,
    )

    fun verify(): Boolean = DevSigner.verify(sender, Canonical.bytes(signedContent()), sig)

    companion object {
        fun create(
            signer: Signer, payload: Map<String, Any?>, msgId: String, lamport: Int,
            type: String = "sos", ttl: Int = 4,
        ): SignedEnvelope {
            val env = SignedEnvelope(signer.nodeId, msgId, lamport, payload, "", type, 1, ttl)
            env.sig = signer.sign(Canonical.bytes(env.signedContent()))
            return env
        }
    }
}

/** A mesh node: Lamport clock + dedup + multi-hop flooding. */
class MeshNode(val nodeId: String) {
    var clock: Int = 0
    val seen: MutableSet<String> = mutableSetOf()
    val inbox: MutableList<Map<String, Any?>> = mutableListOf()

    /** Returns an envelope to re-broadcast (ttl-1), or null (dropped/duplicate/invalid). */
    fun receive(env: SignedEnvelope): SignedEnvelope? {
        if (!env.verify()) return null                       // tamper/forgery -> drop
        if (env.msgId in seen) return null                   // duplicate -> drop
        seen.add(env.msgId)
        clock = maxOf(clock, env.lamport) + 1                // Lamport update
        inbox.add(env.payload)
        if (env.ttl > 0) return env.copy(ttl = env.ttl - 1)  // sig still valid (ttl not signed)
        return null
    }
}
